salt/transport/zeromq.py

```diff
@@ -380,8 +380,9 @@ def __del__(self):
     def connect(self):
         if not self.auth.authenticated:
             yield self.auth.authenticate()
-        self.publish_port = self.auth.creds['publish_port']
-        self._socket.connect(self.master_pub)
+        if self.opts.get('detect_mode', False) is False:
+            self.publish_port = self.auth.creds['publish_port']
+            self._socket.connect(self.master_pub)
```

Comment:
Without this change, if the master is set to `zeromq` and the minion is set to detect, you will get this error:
```
Traceback (most recent call last):
  File "/usr/lib/python2.7/site-packages/tornado/ioloop.py", line 604, in _run_callback
    ret = callback()
  File "/usr/lib/python2.7/site-packages/tornado/stack_context.py", line 275, in null_wrapper
    return fn(*args, **kwargs)
  File "/usr/lib/python2.7/site-packages/tornado/ioloop.py", line 619, in <lambda>
    self.add_future(ret, lambda f: f.result())
  File "/usr/lib/python2.7/site-packages/tornado/concurrent.py", line 237, in result
    raise_exc_info(self._exc_info)
  File "/usr/lib/python2.7/site-packages/tornado/gen.py", line 1024, in run
    yielded = self.gen.send(value)
  File "/usr/lib/python2.7/site-packages/salt/crypt.py", line 505, in _authenticate
    self._authenticate_future.set_exception(error)
  File "/usr/lib/python2.7/site-packages/tornado/concurrent.py", line 283, in set_exception
    getattr(exception, '__traceback__', None)))
  File "/usr/lib/python2.7/site-packages/tornado/concurrent.py", line 306, in set_exc_info
    self._set_done()
  File "/usr/lib/python2.7/site-packages/tornado/concurrent.py", line 320, in _set_done
    for cb in self._callbacks:
TypeError: 'NoneType' object is not iterable
```
The error does not appear if the master is set to `tcp`.